app/javascript/mastodon/reducers/modal.ts

```diff
@@ -0,0 +1,91 @@
+import { Record as ImmutableRecord, Stack } from 'immutable';
+
+import type { PayloadAction } from '@reduxjs/toolkit';
+
+import { COMPOSE_UPLOAD_CHANGE_SUCCESS } from '../actions/compose';
+import type { ModalType } from '../actions/modal';
+import { openModal, closeModal } from '../actions/modal';
+import { TIMELINE_DELETE } from '../actions/timelines';
+
+type ModalProps = Record<string, unknown>;
+interface Modal {
+  modalType: ModalType;
+  modalProps: ModalProps;
+}
+
+const Modal = ImmutableRecord<Modal>({
+  modalType: 'ACTIONS',
+  modalProps: ImmutableRecord({})(),
+});
+
+interface ModalState {
+  ignoreFocus: boolean;
+  stack: Stack<ImmutableRecord<Modal>>;
+}
+
+const initialState = ImmutableRecord<ModalState>({
+  ignoreFocus: false,
+  stack: Stack(),
+})();
+
+type PopModal = (
+  state: typeof initialState,
+  option: { modalType: ModalType | undefined; ignoreFocus: boolean }
+) => typeof initialState;
+const popModal: PopModal = (state, { modalType, ignoreFocus }) => {
+  if (
+    modalType === undefined ||
+    modalType === state.get('stack').get(0)?.get('modalType')
+  ) {
+    return state
+      .set('ignoreFocus', !!ignoreFocus)
+      .update('stack', (stack) => stack.shift());
+  } else {
+    return state;
+  }
+};
+
+type PushModal = (
+  state: typeof initialState,
+  modalType: ModalType,
+  modalProps: ModalProps
+) => typeof initialState;
```

Comment:
This is not idiomatic, in this case you probably want to use `function()` with the correct argument & return types.

Both are pretty much the same (except the arrow function vs function).

Some prefer to extract the types, so it is easier to re-use them elsewhere without having to use `ReturnType<typeof myFunction>`, but I do not think we have a preferred syntax (yet) in the codebase.

Reply:
I see. In this case I don't think this type would be reused elsewhere as it's pretty much an internal function. Therefore it'd make more sense to me to not name the type. That being said, I'm not that familiar with Typescript, so I'll let you decide on that.

Reply:
Often in Typescript, the first argument is an object.
This is similar to keyword argument in Ruby.
And JavaScript can [Destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) in function arguments.

In this case, there are some style for type annotation for function.

```typescript
type SomeFunction = (args: {
  str: string;
}) => string
const someFunction: SomeFunction = ({ str }) => {
  return(`some ${str}`)
}
```

```typescript
type SomeFunctionArg = {
  str: string;
};
const someFunction = ({ str }: SomeFunctionArg ): string => {
  return(`some ${str}`)
}
```

```typescript
const someFunction = ({ str }: {
  str: string;
}): string => {
  return(`some ${str}`)
}
```

IMO, I seems that 1st style are most simply for me.
Because typed variables accepts function object or not.

Reply:
I agree that it does not have to be the Arrow style.
However, when we use both of Arrow and `function` styles, a decision must be required about which to use **every time**.

If we use only an arrow style, there is no need to consider about it.
Tthis is an important advantage for newbie contributor friendly, I think.

In addition, I favourite that there is no confusion caused by `this` when use the arrow style.

I want to make these codes readable, mantainable, improvable for everyone even if not javascript specialist.
For it, We have to reduce confusion and cost of dicision as a possible.

Reply:
I rewrite type annotations into arrow function definitions.

@renchap Could you merge this PR?